File: ios/Classes/Utils/ArtworkUtils.swift
/*
 * ArtworkUtils.swift
 *
 * Copyright (c) 2020-2025 Ilia Chirkunov <contact@cheebeez.com>
 *
 * This source code is licensed under the CC BY-NC-SA 4.0.
 * See https://creativecommons.org/licenses/by-nc-sa/4.0/
 */

import UIKit 
import Foundation

struct ArtworkUtils {
    /// Downloads an image from a URL.
    static func downloadImage(from urlString: String?) async -> UIImage? {
        guard let actualUrlString = urlString,
              let url = URL(string: actualUrlString) else { return nil }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            return UIImage(data: data)
        } catch {
            print("ArtworkUtils: Failed to download image from \(actualUrlString): \(error)")
            return nil
        }
    }

    /// Fetches an artwork URL from iTunes API for the given artist and track.
    static func parseArtworkFromItunes(artist: String, track: String) async -> String {
        guard let term = (artist + " - " + track).addingPercentEncoding(withAllowedCharacters: .alphanumerics),
              let url = URL(string: "https://itunes.apple.com/search?term=\(term)&limit=1")
        else { return "" }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            guard let dict = try JSONSerialization.jsonObject(with: data, options: .allowFragments) as? [String: Any],
                  let resultCount = dict["resultCount"] as? Int, resultCount > 0,
                  let results = dict["results"] as? [[String: Any]], !results.isEmpty,
                  let artworkUrl30 = results[0]["artworkUrl30"] as? String
            else { return "" }

            let highResArtworkUrl = artworkUrl30.replacingOccurrences(of: "30x30bb", with: "500x500bb")
            return highResArtworkUrl
        } catch {
            print("ArtworkUtils: Failed to parse iTunes artwork for \(artist) - \(track): \(error)")
            return ""
        }
    }
}

File: android/src/main/kotlin/com/cheebeez/radio_player/utils/ArtworkUtils.kt
/*
 * ArtworkUtils.kt
 *
 * Copyright (c) 2020-2025 Ilia Chirkunov <contact@cheebeez.com>
 *
 * This source code is licensed under the CC BY-NC-SA 4.0.
 * See https://creativecommons.org/licenses/by-nc-sa/4.0/
 */

package com.cheebeez.radio_player

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.URL
import java.net.URLEncoder

/// Downloads an image from a URL.
suspend fun downloadImage(value: String?): ByteArray? {
    if (value == null) return null

    return try {
        withContext(Dispatchers.IO) {
            val url = URL(value)
            url.openStream().use { inputStream ->
                inputStream.readBytes()
            }
        }
    } catch (e: Throwable) {
        Log.e("ArtworkUtils", "Error downloading image: ${e.message}") 
        null
    }
}

/// Fetches an artwork URL from iTunes API for the given artist and track.
suspend fun parseArtworkFromItunes(artist: String, track: String): String {
    if (artist.isBlank() && track.isBlank()) return ""

    return try {
        val term = URLEncoder.encode("$artist - $track".trim(), "utf-8")

        val response = withContext(Dispatchers.IO) {
            URL("https://itunes.apple.com/search?term=$term&media=music&entity=song&limit=1").readText()
        }

        val jsonObject = JSONObject(response)
        var artwork = ""

        if (jsonObject.getInt("resultCount") > 0) {
            val artworkUrl30: String = jsonObject.getJSONArray("results").getJSONObject(0).getString("artworkUrl30")
            artwork = artworkUrl30.replace("30x30bb","500x500bb")
        }

        artwork
    } catch (e: Throwable) {
        Log.e("ArtworkUtils", "Error parsing artwork from iTunes: ${e.message}")
        ""
    }
}
